### PR TITLE
Documentation improvements

### PR DESCRIPTION
**Pull request by zachmullen**

### CODE REVIEW

**Comment by zachmullen** [comment]
@kotfic PTAL, this is the follow up PR on top of the cleaning / overhaul of the documentation. This one is primarily focused on docker_run documentation, though it also has the effect of fixing our sphinx autodocs to be useful, along with a bunch of sphinx bugfixes in our docstrings.

File: docs/builtin-plugins.rst [hunk 1]
```diff
@@ -198,7 +201,99 @@ and start reading. Below is an example python entry point:
...
         fp.read() # This will be reading the files contents
 
+.. _docker-progress:
+
+Streaming progress reporting from Docker tasks to Girder jobs
+-------------------------------------------------------------
+
+The :py:class:`girder_worker.docker.transforms.girder.ProgressPipe` class can be used to facilitate
+streaming real-time progress reporting from a docker task to its associated Girder job. It uses
+a named pipe to provide a simplistic interface within the container that is usable from any
```
Comment (kotfic) [inline docs/builtin-plugins.rst]: simplistic => simple

File: docs/builtin-plugins.rst [hunk 2]
```diff
@@ -198,7 +201,99 @@ and start reading. Below is an example python entry point:
...
+
+Attaching intermediate / optional artifacts to Girder jobs
+----------------------------------------------------------
+
+It's often useful for debugging/tracing or algorithm analysis to be able to inspect intermediate
+outputs or other artifacts produced during execution of a task, even (perhaps especially) if the
+task fails. These artifacts differ from normal output transforms that upload files to Girder in
+two ways. Firstly, they are optional; if the specified file or directory does not exist, it does
+not cause any errors. This allows docker image authors to choose either at build time or runtime
+whether or not to emit artifacts. Secondly, the emitted files are attached to the job document
```
Comment (kotfic) [inline docs/builtin-plugins.rst]: consider: 

"This allows docker image authors to choose either at build time or runtime whether or not to create and upload artifacts. Secondly, the artifact files are attached to the job document..."

File: docs/builtin-plugins.rst [hunk 3]
```diff
@@ -198,7 +201,99 @@ and start reading. Below is an example python entry point:
...
+        girder_result_hooks=[
+            GirderUploadVolumePathJobArtifact(artifacts)
+        ])
+
+Note that you can write to this path inside your container and make it either a directory or
+a single file. If it's a directory, all files within the directory will be uploaded and attached
+to the job as artifacts. This operation is not recursive, i.e. it will not upload anything under
+subdirectories of the top level directory.
+
+It's often useful to upload any emitted artifact files even if the ``docker_run`` task failed.
```
Comment (kotfic) [inline docs/builtin-plugins.rst]: consider deleting "emitted"

File: girder_worker/__init__.py [hunk 4]
```diff
@@ -35,14 +35,18 @@
 
 @add_metaclass(abc.ABCMeta)
 class GirderWorkerPluginABC(object):
-    """ """
+    """
+    Abstract base class for Girder Worker plugins. Plugins must descend from this
+    class; see the :ref:`plugins` section for information about how to do so.
+    """
```
Comment (kotfic) [inline girder_worker/__init__.py]: considering deleting "about how to do so."

File: girder_worker/task.py [hunk 5]
```diff
@@ -23,7 +23,11 @@ def job(self):
 
 
 class Task(celery.Task):
-    """Girder Worker Task object"""
+    """
+    Girder Worker Task object. Tasks defined by plugins must be subclasses of this class,
+    however you will not need to reference it yourself typically, as it will be automatically
```
Comment (kotfic) [inline girder_worker/task.py]: consider "however, typically, you will not need to reference it yourself..."

**Comment by zachmullen** [comment]
Thanks for the review, all good suggestions. Will merge once CI returns

**Pull request merged**